### PR TITLE
support/db: Add RowStruct() to BatchInsertBuilder

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Close https://github.com/stellar/go/issues/2578

* Add `RowStruct()` to BatchInsertBuilder which allows you to pass in a struct representation of a row instead of a `map[string]interface{}` value.
* Update transaction builder to use `RowStruct()` instead of `Row()`

### Why

`map[string]interface{}` isn't an ideal representation for a row because there aren't any guarantees provided by the type system that two `map[string]interface{}` values have the same length and the same set of keys.

In https://github.com/stellar/go/pull/2562 , we had to fix an issue where a `TransactionBatchInsertBuilder` was calling `Row()` on two `map[string]interface{}` values which were not consistent with each other. This manifested in a runtime error which blocked ingestion.

If we can pass `history.Transaction` instances to `Row()` instead of a `map[string]interface{}` values, we can know that all row values will have the same number of columns and the same set of column names. This would make ingestion more robust to the kind of bug we had to fix in https://github.com/stellar/go/pull/2562 .

### Known limitations

[N/A]

### CODE REVIEW

File: services/horizon/internal/db2/history/transaction_batch_insert_builder.go [hunk 1]
```diff
@@ -130,61 +130,88 @@ func memo(transaction io.LedgerTransaction) null.String {
 	return null.NewString(value, valid)
 }
 
-func transactionToMap(transaction io.LedgerTransaction, sequence uint32) (map[string]interface{}, error) {
+type transactionRow struct {
```
Comment (bartekn) [inline services/horizon/internal/db2/history/transaction_batch_insert_builder.go:133]: I wonder if we can reuse `db2.history.Transaction` for inserts? This would require more work (ex. in `resourceadapter` package and figuring out `sqx.StringArray`) but would remove code duplication: when adding/changing a new field we need to update more than one place.

File: services/horizon/internal/db2/history/transaction_batch_insert_builder.go [hunk 2]
```diff
@@ -43,12 +43,12 @@ func (q *Q) NewTransactionBatchInsertBuilder(maxBatchSize int) TransactionBatchI
 
 // Add adds a new transaction to the batch
 func (i *transactionBatchInsertBuilder) Add(transaction io.LedgerTransaction, sequence uint32) error {
```
Comment (bartekn) [inline services/horizon/internal/db2/history/transaction_batch_insert_builder.go:45]: Definitely not in a scope in this PR but maybe you want to work on this. This is an issue I introduced, I think using `io.LedgerTransaction` here breaks a separation of concerns because if (for whatever reason) we decide to insert transactions from outside `ingest` package, we would need to import `ingest/io` to be able to do so. Also, whoever is working on `db2/history` should not need to understand how `io.LedgerTransaction` work.

In short: this method should accept `db2/history.Transaction` (or another struct used for inserts) instead of `io.LedgerTransaction`.

File: services/horizon/internal/db2/history/transaction_batch_insert_builder.go (hunk 1 above)
Comment (tamirms) [reply services/horizon/internal/db2/history/transaction_batch_insert_builder.go:133]: I looked into this and I couldn't find a clean way of doing it. The SQL query to read transactions joins from the `history_ledgers` table to get the ledger close time which is written to the `LedgerCloseTime` attribute of the `Transaction` struct. When inserting transactions we want to exclude `LedgerCloseTime`.

Comment (bartekn) [reply services/horizon/internal/db2/history/transaction_batch_insert_builder.go:133]: What about something like this:
```go
// Used for inserts
type TransactionWithoutLedger struct {
	TotalOrderID
	TransactionHash      string      `db:"transaction_hash"`
	LedgerSequence       int32       `db:"ledger_sequence"`
	// ...
}

// Transaction is a row of data from the `history_transactions` table
type Transaction struct {
	LedgerCloseTime time.Time `db:"ledger_close_time"`
	TransactionWithoutLedger
}
```
I wasn't sure if `db` tags would be correctly handled but I checked it locally and it works.

Sorry for slowing down the merge but I think it would be really great to avoid duplicating transaction struct.

File: services/horizon/internal/db2/history/transaction_batch_insert_builder.go [hunk 3]
```diff
@@ -130,61 +150,97 @@ func memo(transaction io.LedgerTransaction) null.String {
...
+	MemoType             string              `db:"memo_type"`
+	Memo                 null.String         `db:"memo"`
+	TimeBounds           pgtype.Int8range    `db:"time_bounds"`
+	CreatedAt            time.Time           `db:"created_at"`
+	UpdatedAt            time.Time           `db:"updated_at"`
+	Successful           bool                `db:"successful"`
+	FeeAccount           null.String         `db:"fee_account"`
+	InnerTransactionHash null.String         `db:"inner_transaction_hash"`
+	NewMaxFee            null.Int            `db:"new_max_fee"`
+	InnerSignatures      pgtype.VarcharArray `db:"inner_signatures"`
```
Comment (bartekn) [inline services/horizon/internal/db2/history/transaction_batch_insert_builder.go:177]: It looks we use only two types from `pgtype`. Do you think we could implement our own type with that implements `driver.Valuer` and `sql.Scanner` interfaces?

I can see two advantages of doing so. First, we actually import/update ~40 packages and we should check them from security perspective. Second, we should stop using `Int8range` for storing time bounds at some point in the future. Time bounds are using `uint64` which is not supported by postgres. We actually have the special case in the code (`maxTime > math.MaxInt64`) that overrides max time in case of an overflow and by doing so we present invalid data.

Comment (tamirms) [reply services/horizon/internal/db2/history/transaction_batch_insert_builder.go]: @bartekn I have updated the PR to remove the dependency on `pgtype`
